gadget/install/partition.go

```diff
@@ -164,58 +164,53 @@ func buildPartitionList(dl *gadget.OnDiskVolume, pv *gadget.LaidOutVolume, opts
 
 	// Write new partition data in named-fields format
 	buf := &bytes.Buffer{}
-	for _, p := range pv.LaidOutStructure {
-		// Make loop var per-iter as we store the pointer in the results
-		p := p
-		if !p.IsPartition() {
+	for _, laidOut := range odv.LaidOutStructure {
+		if !laidOut.GadgetStructure.IsPartition() {
 			continue
 		}
 
 		pIndex++
-		s := p.VolumeStructure
 
 		// Skip partitions that are already in the volume
-		startInSectors := uint64(p.StartOffset) / sectorSize
+		startInSectors := uint64(laidOut.StartOffset) / sectorSize
 		if seen[startInSectors] {
 			continue
 		}
 
 		// Only allow creating certain partitions, namely the ubuntu-* roles
-		if !opts.CreateAllMissingPartitions && !gadget.IsCreatableAtInstall(p.VolumeStructure) {
-			return nil, nil, fmt.Errorf("cannot create partition %s", p)
+		if !opts.CreateAllMissingPartitions && !gadget.IsCreatableAtInstall(laidOut.GadgetStructure) {
+			return nil, nil, fmt.Errorf("cannot create partition %s", laidOut)
 		}
 
 		// Check if the data partition should be expanded
-		newSizeInSectors := uint64(s.Size) / sectorSize
-		if s.Role == gadget.SystemData && canExpandData && startInSectors+newSizeInSectors < dl.UsableSectorsEnd {
+		newSizeInSectors := uint64(laidOut.Size) / sectorSize
+		if laidOut.GadgetStructure.Role == gadget.SystemData && canExpandData && startInSectors+newSizeInSectors < dl.UsableSectorsEnd {
 			// note that if startInSectors + newSizeInSectors == dl.UsableSectorEnd
 			// then we won't hit this branch, but it would be redundant anyways
 			newSizeInSectors = dl.UsableSectorsEnd - startInSectors
 		}
 
-		ptype := partitionType(dl.Schema, p.Type)
+		ptype := partitionType(dl.Schema, laidOut.Type)
 
 		// synthesize the node name and on disk structure
 		node := deviceName(dl.Device, pIndex)
-		ps := onDiskAndLaidoutStructure{
-			onDisk: &gadget.OnDiskStructure{
-				Name:        p.Name,
-				Label:       p.Label,
-				Type:        p.Type,
-				Filesystem:  p.Filesystem,
-				StartOffset: p.StartOffset,
-				Node:        node,
-				DiskIndex:   pIndex,
-				Size:        quantity.Size(newSizeInSectors * sectorSize),
-			},
-			laidOut: &p,
+		// TODO fill in construction of LaidOutStructure instead?
+		laidOut.OnDiskStructure = gadget.OnDiskStructure{
+			Name:        laidOut.Name,
+			Label:       laidOut.Label,
+			Type:        laidOut.Type,
+			Filesystem:  laidOut.Filesystem,
+			StartOffset: laidOut.StartOffset,
+			Node:        node,
+			DiskIndex:   pIndex,
+			Size:        quantity.Size(newSizeInSectors * sectorSize),
 		}
```

Comment:
this definitely read strangely? also isn't some of this already filled? where is odv coming from?